### PR TITLE
#90757 Implement a messagelock duration calculator

### PR DESCRIPTION
**Pull request by tabman83**

### CODE REVIEW

File: src/CaptainHook.EventReaderService/MessageLockDurationCalculator.cs [hunk 1]
```diff
@@ -0,0 +1,27 @@
...
+            {
+                throw new ArgumentNullException(nameof(retrySleepDurations));
+            }
+
+            var result =
+                httpTimeout.TotalSeconds * (retrySleepDurations.Length + 1) +
+                retrySleepDurations.Sum(x => x.TotalSeconds) +
+                TimeAllowance.TotalSeconds;
+
+            return result > 300 ? 300 : (int)result;
```
Comment (chomick) [inline src/CaptainHook.EventReaderService/MessageLockDurationCalculator.cs:24]: is that a service bus limit? maybe you should have it as const?

Comment (tabman83) [reply src/CaptainHook.EventReaderService/MessageLockDurationCalculator.cs:24]: It is a Service Bus limit, yes.

Comment (tabman83) [reply src/CaptainHook.EventReaderService/MessageLockDurationCalculator.cs]: Changed with constant

**Review by mszymeshop — APPROVED** [review]

File: src/CaptainHook.EventReaderService/MessageLockDurationCalculator.cs [hunk 2]
```diff
@@ -0,0 +1,30 @@
+﻿using System;
+using System.Linq;
+
+namespace CaptainHook.EventReaderService
+{
+    public class MessageLockDurationCalculator : IMessageLockDurationCalculator
+    {
+        private static readonly TimeSpan TimeAllowance = TimeSpan.FromSeconds(5);
```
[the remark this answers is not in the document]
Comment (tabman83) [reply src/CaptainHook.EventReaderService/MessageLockDurationCalculator.cs:8]: Changed

File: src/CaptainHook.EventReaderService/IMessageLockDurationCalculator.cs [hunk 3]
```diff
@@ -0,0 +1,15 @@
...
+namespace CaptainHook.EventReaderService
+{
+    public interface IMessageLockDurationCalculator
+    {
+        /// <summary>
+        /// Calculates the message lock duration in seconds given the http timeout and retry durations
+        /// </summary>
+        /// <param name="httpTimeout"></param>
+        /// <param name="retrySleepDurations"></param>
+        /// <returns></returns>
```
[the remark this answers is not in the document]
Comment (tabman83) [reply src/CaptainHook.EventReaderService/IMessageLockDurationCalculator.cs:12]: Changed

File: src/CaptainHook.EventReaderService/MessageLockDurationCalculator.cs [hunk 4]
```diff
@@ -0,0 +1,30 @@
...
+            {
+                throw new ArgumentNullException(nameof(retrySleepDurations));
+            }
+
+            var result =
+                httpTimeout.TotalSeconds * (retrySleepDurations.Length + 1) +
+                retrySleepDurations.Sum(x => x.TotalSeconds) +
+                TimeAllowance.TotalSeconds;
+
+            return result > MaxAllowedMessageLockDurationInSeconds ?
```
[the remark this answers is not in the document]
Comment (tabman83) [reply src/CaptainHook.EventReaderService/MessageLockDurationCalculator.cs:25]: Changed with Math.Min() instead as discussed over Slack

**Pull request merged**